middleware.js

```diff
@@ -12,11 +13,43 @@ function webpackHotMiddleware(compiler, opts) {
   var eventStream = createEventStream(opts.heartbeat);
   var latestStats = null;
 
-  compiler.plugin("compile", function() {
-    latestStats = null;
-    if (opts.log) opts.log("webpack building...");
-    eventStream.publish({action: "building"});
+  var compilers = compiler.compilers || [compiler];
+  compilers.forEach(function(compiler, index) {
+    compiler.id = "__WEBPACK_COMPILER_" + index + "__";
+    compiler.plugin("compilation", function (compilation, params) {
+      var hotUpdateChunkTemplate = compilation.hotUpdateChunkTemplate;
+      if(!hotUpdateChunkTemplate) return;
+
+      compilation.mainTemplate.plugin("require-extensions", function(source) {
```

Comment:
Are there any docs or references you could link to in a comment for how this stuff works? From reading it I can see roughly what it's doing, but it looks like it's using some pretty gnarly internal webpack APIs?